kinto/core/storage/postgresql/__init__.py

```diff
@@ -647,21 +642,24 @@ def get_all(self, collection_id, parent_id, filters=None, sorting=None,
         if pagination_rules:
             sql, holders = self._format_pagination(pagination_rules, id_field,
                                                    modified_field)
-            safeholders['pagination_rules'] = 'WHERE {}'.format(sql)
+            safeholders['pagination_rules'] = '{} AND'.format(sql)
             placeholders.update(**holders)
+        else:
+            safeholders['pagination_rules'] = ''
 
         # Limit the number of results (pagination).
         limit = min(self._max_fetch_size, limit) if limit else self._max_fetch_size
         placeholders['pagination_limit'] = limit
 
         with self.client.connect(readonly=True) as conn:
-            result = conn.execute(query.format_map(safeholders), placeholders)
+            result = conn.execute(select_query.format_map(safeholders), placeholders)
             retrieved = result.fetchmany(self._max_fetch_size)
 
-        if len(retrieved) == 0:
-            return [], 0
+            if len(retrieved) == 0:
+                return [], 0
```

Comment:
There might be some further work to do here. 
This early exit allows us to being able to *skip* the COUNT query since the SELECT query yielded 0 results anyway. 

The reason for why I didn't use the exact filter for the COUNT and the SELECT was because of this test:
https://github.com/peterbe/kinto/blob/51c7bf9ec9eb1f89d7e7d3a015429f2bc525551d/kinto/core/storage/testing.py#L1217-L1218
(there might be more like this!)
In that test it does a `get_all(include_deleted=True, ...)` and it expects the number of records to be different from the total count. 
Isn't that strange? If you `include_deleted=True` why should the total count be always filter out the deleted records??

If that test is wrong, I can rewrite the two queries so that the SELECT and COUNT uses the exact same WHERE clause. Then, if we do that we can do the COUNT first (slightly faster since it's just an integer) and if it's 0 we can skip the SELECT and just return an empty list.